File: algorithms/pathSum/pathSum.II.js
/****
 * Solution of https://leetcode.com/problems/path-sum-ii/
 * @author: Jerry Zou
 * @date: 2015.6.9
 **/

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
var pathSum = function(root, sum) {
    var result = [],
        target = sum,
        path = [];

    function simpleClone(arr) {
        var clone = [];
        for (var i = 0; i < arr.length; i++) clone.push(arr[i]);
        return clone;
    }

    function partialPathSum(root, current) {
        if (!root) {
            return;
        }

        current += root.val;

        if (current === target && !(root.left || root.right)) {
            path.push(root.val);
            result.push(simpleClone(path));
            path.pop();
        } else if (root.left || root.right){
            path.push(root.val);
            partialPathSum(root.left, current);
            partialPathSum(root.right, current);
            path.pop();
        } 
    }

    partialPathSum(root, 0);
    return result;
};

File: algorithms/cpp/pathSum/pathSum.II.cpp
// Source : https://oj.leetcode.com/problems/path-sum-ii/
// Author : Hao Chen
// Date   : 2014-07-01

/********************************************************************************** 
* 
* Given a binary tree and a sum, find all root-to-leaf paths where each path's sum equals the given sum.
* 
* For example:
* Given the below binary tree and sum = 22,
* 
*               5
*              / \
*             4   8
*            /   / \
*           11  13  4
*          /  \    / \
*         7    2  5   1
* 
* return
* 
* [
*    [5,4,11,2],
*    [5,8,4,5]
* ]
* 
*               
**********************************************************************************/

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > pathSum(TreeNode *root, int sum) {
        vector<vector<int> > result;
        vector<int> v;
        generatePathSum(root, sum, 0, v, result);
        return result;
    }

    void generatePathSum(TreeNode *root, int sum, int s, vector<int> v, vector<vector<int> >& result) {
        if (root==NULL) return ;

        s += root->val;
        v.push_back(root->val);

        if ( root->left==NULL && root->right==NULL) {
            if (s == sum) result.push_back(v);
            return;
        }

        generatePathSum(root->left, sum, s, v, result);
        generatePathSum(root->right, sum, s, v, result);
    }
};
